design-proposals/amd-sev.md

```diff
@@ -0,0 +1,245 @@
+
+### Step #1 Validation of the platform
+
+The guest owner should get the assets listed bellow from the platform owner:
+
+- the `Platform Diffie-Hellman (PDH)` key exported from the SEV hardware
+- the complete key chain (up to the `root of trust`) used to sign the `PDH`
+- the OVMF image
+
+On a node where KubeVirt is running, the privileged `virt-handler` pod can export the required assets in advance to a `kubevirt-private` directory. Whenever a new VM requiring SEV is launched, the directory can be mounted to the `virt-launcher` pod in `read-only` mode. A user can then invoke the `virtctl` tool to download the certificate chain for validation (for that an additional command will be needed, e.g. `virtctl fetchcertchain ...`).
+
+In order to ensure that the platform is not compromised the guest owner needs to validate the `PDH` key using the provided chain. For that `sev-tool --validate_cert_chain ...` can be used (AMD root certificate should be downloaded from the web for verification). Additionally the UEFI image should be considered 'trustable' by the guest owner.
+
+### Step #2 Establishing secure channel with the platform
+
+To establish a secure communication channel with the platform a guest owner needs to provide a launch blob with the parameters of the session and the certificate derived from `PDH`. This is currently not in the scope of KubeVirt as it requires the guest owner to expose the secrets. It should be performed 'offline' by the guest owner (e.g. by using `sev-tool --generate_launch_blob ...`).
+
+The blob and certificate can be passed via `session` and `dhCert` parameters of the libvirt domain xml. This however brings a limitation: the guest owner needs to know in advance what node the VM will be scheduled on (which is usually not the case). Therefore additional handling is required in order to mitigate that in KubeVirt.
+
+Initially the guest owner can provide empty `dhCert` and `session` parameters in the VMI spec:
+
+```yaml
+spec:
+  domain:
+    launchSecurity:
+      sev:
+        dhCert: ""
+        session: ""
+```
+
+The implementation in `virt-launcher` in turn can 'postpone' the creation of the domain until the parameters are specified. Sidecar hooks functionality can be used to modify the domxml after a VMI has been scheduled. New `virtctl` commands (e.g. `virtctl launchblob ...`) can be introduced to support the process and to trigger propagation of the parameters.
+
+To run a SEV VM on a specific node the parameters can be specified directly in the VMI spec:
+
+```yaml
+spec:
+  nodeSelector:
+    kubernetes.io/hostname: "node01"
```

Comment:
This is also can be done with a limited success, as there may be other requirements that will prevent the scheduler from placing the VMI on this node.

Reply:
Right. There is no guarantee here.

Reply:
Here a user needs to somehow get the platform certificates from node01 upfront. AFAIK that's all what is required for providing the needed parameters `dhCert` and `session`.